src/gafaelfawr/models/history.py

```diff
@@ -44,3 +159,146 @@ class TokenChange(Enum):
     revoke = "revoke"
     expire = "expire"
     edit = "edit"
+
+
+class TokenChangeHistoryEntry(BaseModel):
+    """A record of a change to a token."""
+
+    token: str = Field(
+        ...,
+        title="Key of the token that was changed",
+        min_length=22,
+        max_length=22,
+    )
+
+    username: str = Field(
+        ...,
+        title="Username of the token",
+        min_length=1,
+        max_length=64,
+    )
+
+    token_type: TokenType = Field(..., title="Type of the token")
+
+    token_name: Optional[str] = Field(
+        None,
+        title="Name of the token",
+        description=(
+            "Only set for tokens of type user. If the name was changed, this"
+            " will be the new name of the token."
+        ),
+    )
+
+    parent: Optional[str] = Field(
+        None, title="Key of parent token of this token"
+    )
+
+    scopes: List[str] = Field(..., title="Scopes of the token")
+
+    service: Optional[str] = Field(
+        None,
+        title="Service to which the token was issued",
+        description="Only set for tokens of type internal.",
+    )
+
+    expires: Optional[datetime] = Field(
+        None,
+        title="Expiration of the token",
+        description=(
+            "If the expiration was changed, this will be the new expiration of"
+            " the token."
+        ),
+    )
+
+    actor: str = Field(
+        ...,
+        title="Username of person making the change",
+        min_length=1,
+        max_length=64,
+    )
+
+    action: TokenChange = Field(..., title="Type of change that was made")
+
+    old_token_name: Optional[str] = Field(
+        None,
+        title="Previous name of the token",
+        description=(
+            "This field will only be present for edit changes to user tokens"
+            " that changed the token name."
+        ),
+    )
+
+    old_scopes: Optional[List[str]] = Field(
+        None,
+        title="Previous scopes of the token",
+        description=(
+            "This field will only be present for edit changes that changed the"
+            " token scopes."
+        ),
+    )
+
+    old_expires: Optional[datetime] = Field(
+        None,
+        title="Previous expiration of the token",
+        description=(
+            "This field will only be present for edit changes that changed the"
+            " expiration of the token."
+        ),
+    )
+
+    ip_address: Optional[str] = Field(
```

Comment:
I tried that and it turned into a morass of type conversions because the database layer wants a string and the corresponding query is either an IP address or a CIDR block. I therefore unwound that and did manual validation of the query and that code seemed simpler, but then didn't leave a comment behind. I've added a comment now to explain that.